src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb

```diff
@@ -559,28 +583,108 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            ' we don't have to compare against all 3 BoundKinds
+            Dim validAddressOfConversionSyntax = operand.Syntax.Kind() = SyntaxKind.AddressOfExpression AndAlso
+                                                 (conversionSyntax.Kind() = SyntaxKind.CTypeExpression OrElse
+                                                  conversionSyntax.Kind() = SyntaxKind.DirectCastExpression OrElse
+                                                  conversionSyntax.Kind() = SyntaxKind.TryCastExpression OrElse
+                                                  conversionSyntax.Kind() = SyntaxKind.ObjectCreationExpression OrElse
+                                                  (conversionSyntax.Kind() = SyntaxKind.AddressOfExpression AndAlso
+                                                   conversionSyntax Is operand.Syntax))
+
+            Dim validLambdaConversionNode = operand.Kind = BoundKind.Lambda OrElse
+                                              operand.Kind = BoundKind.QueryLambda OrElse
+                                              operand.Kind = BoundKind.UnboundLambda
+
+            Dim validTargetType = targetType.IsDelegateType()
+
+            Return validTargetType AndAlso (validAddressOfConversionSyntax OrElse validLambdaConversionNode)
+        End Function
+
+        ''' <summary>
+        ''' Creates the Lazy IOperation from a delegate creation operand or a bound conversion operand, handling when the conversion
+        ''' is actually a delegate creation.
+        ''' </summary>
+        Private Function CreateConversionOperand(operand As BoundNode, conversionKind As ConversionKind, conversionSyntax As SyntaxNode, targetType As TypeSymbol
+                                                   ) As (methodSymbol As MethodSymbol, operation As Lazy(Of IOperation), isDelegateCreation As Boolean)
+            If (conversionKind And VisualBasic.ConversionKind.UserDefined) = VisualBasic.ConversionKind.UserDefined Then
+                Dim userDefinedConversion As BoundUserDefinedConversion = DirectCast(operand, BoundUserDefinedConversion)
+                Return (userDefinedConversion.Call.Method, New Lazy(Of IOperation)(Function() Create(userDefinedConversion.Operand)), isDelegateCreation:=False)
+            ElseIf IsDelegateCreation(conversionKind, conversionSyntax, operand, targetType) Then
+                Dim methodSymbol As MethodSymbol = Nothing
+                Dim operandLazy As Lazy(Of IOperation)
+                If operand.Kind = BoundKind.DelegateCreationExpression Then
+                    ' If the child is a BoundDelegateCreationExpression, we don't want to generate a nested IDelegateCreationExpression.
+                    ' So, the operand for the conversion will be the child of the BoundDelegateCreationExpression.
+                    operandLazy = New Lazy(Of IOperation)(Function() CreateBoundDelegateCreationExpressionChildOperation(DirectCast(operand, BoundDelegateCreationExpression)))
```

Comment:
I don't follow this. Why do we need to avoid a nested `IDelegateCreationExpression` if the operand is also a `BoundDelegateCreationExpression`? From the tests, it looks like there are still various cases where we get nested instances.

Reply:
I'll add more to the comment, but the problem is that for `Dim x = New Action(AddressOf M2)`, this should be semantically equivalent to `Dim x = AddressOf M2`, but we generate both the BoundConversion and a BoundDelegateCreationExpression. So we need to compensate for that. I have tests to make sure that `Dim x = New Action(CType(AddressOf M2, Action))` is two nested IDelegateCreationExpressions, as we would expect.

---
In reply to: [143330842](https://github.com/dotnet/roslyn/pull/22056#discussion_r143330842) [](ancestors = 143330842)